### PR TITLE
Delete and rename tables

### LINKED ISSUE #183: DB function to rename tables
**Problem**

Users might want to rename tables. We don't currently support this.

**Proposed solution**

A function that handles renaming of tables in the database, including renaming references to them (foreign keys), etc.

**Additional context**

This should be in the `db` module.

### LINKED ISSUE #186: DB function to handle deleting tables
**Problem**

Users might want to delete tables. We don't currently support this.

**Proposed solution**

A function that handles deleting of tables in the database. We should raise an error if there is anything outside of the table referencing the table.

**Additional context**

This should be in the `db` module.

### PR DESCRIPTION
**Pull request by eito-fis**
Fixes #183, Fixes #186

Adds the `delete_table()` and `rename_table()` functions to `db.tables`.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/tables.py [hunk 1]
```diff
@@ -52,6 +54,19 @@ def create_mathesar_table(name, schema, columns_, engine, metadata=None):
     return table
 
 
+def delete_table(name, schema, engine):
```
Comment (kgodey) [inline db/tables.py:57]: How does this handle references to this table from other tables? e.g. another table might have foreign keys pointing to this one.

File: db/tables.py [hunk 2]
```diff
@@ -52,6 +54,19 @@ def create_mathesar_table(name, schema, columns_, engine, metadata=None):
...
+    table = reflect_table(name, schema, engine)
+    table.drop()
+
+
+def rename_table(name, schema, engine, rename_to):
```
Comment (kgodey) [inline db/tables.py:62]: Same question about handling references to this table from other tables.

**Comment by eito-fis** [comment]
Updated tests to handle foreign key cases, and also added the `cascade` and `if_exists` options to the `delete_table` function.

File: db/tables.py [hunk 3]
```diff
@@ -52,6 +55,42 @@ def create_mathesar_table(name, schema, columns_, engine, metadata=None):
...
+@compiler.compiles(DropTableCascade, "postgresql")
+def compile_drop_table(element, compiler, **_):
+    expression = compiler.visit_drop_table(element)
+    if element.cascade:
+        return expression + " CASCADE"
+    else:
+        return expression
+
+
+def delete_table(name, schema, engine, cascade=False, if_exists=False):
```
Comment (kgodey) [inline db/tables.py:74]: If `cascade` is `False`, can you return a list of related items just like with schemas?

**Review by kgodey — CHANGES_REQUESTED** [review]
Looks good overall, just one question. Also, since this is entirely `db` focused, I'd like @mathemancer to look at it quickly before merge.

**Review by mathemancer — COMMENTED** [review]
Similarly to schemata, I think we should have a separate endpoint at something like:
```
/api/v0/tables/<table_id>/dependencies/
```
That would return a paginated set of dependent objects.  There could be many, and I think the pagination would be a good idea.  I suggest trying to use the `pg_depends` table (view?) to crawl the dependency graph.  The reason is that it would be highly undesirable to list some but not all dependent objects, since this could give the user false confidence, and result in accidental deletion of objects they don't want to get rid of.

`CASCADE` is an extremely aggressive flag, and care should be taken here.  I.e., `CASCADE` will follow the dependency graph as far as it goes (and as far as permissions allow).

**Comment by kgodey** [comment]
ll create a separate issue for the list of dependencies and corresponding API. For now, @eito-fis, please update the error message to be generic as @mathemancer suggested in the other issue.

**Comment by eito-fis** [comment]
Similar update to the schema PR, just modified `delete_table` to raise the original `DependentObjectStillExists` error.

**Review by kgodey — APPROVED** [review]

**Comment by kgodey** [comment]
I'm going to merge this since @mathemancer didn't request changes.

**Pull request merged**